lib/update/phase_coredns.go

```diff
@@ -0,0 +1,133 @@
+/*
+Copyright 2018 Gravitational, Inc.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package update
+
+import (
+	"context"
+
+	"github.com/gravitational/gravity/lib/storage"
+	"github.com/gravitational/rigging"
+	"github.com/gravitational/trace"
+	log "github.com/sirupsen/logrus"
+	rbacv1 "k8s.io/api/rbac/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/client-go/kubernetes"
+)
+
+const (
+	corednsResourceName = "gravity:coredns"
+)
+
+func (r phaseBuilder) corednsPhase(leadMaster storage.Server) *phase {
+	phase := root(phase{
+		ID:          "coredns",
+		Description: "Provision coredns resources",
+		Executor:    coredns,
+		Data: &storage.OperationPhaseData{
+			Server: &leadMaster,
+		},
+	})
+	return &phase
+}
+
+type updatePhaseCoreDNS struct {
+	log.FieldLogger
+	kubernetesOperation
+}
+
+// NewPhaseCoreDNS create an upgrade phase to add coredns rbac permissions
+// HACK: This phase needs to run before updating planet, as coredns is embedded in planet
+// and will come online as planet restarts
+func NewPhaseCoreDNS(c FSMConfig, plan storage.OperationPlan, phase storage.OperationPhase) (*updatePhaseCoreDNS, error) {
+	op, err := newKubernetesOperation(c, plan, phase)
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+	return &updatePhaseCoreDNS{
+		kubernetesOperation: *op,
+		FieldLogger: log.WithFields(log.Fields{
+			"phase": phase.ID,
+		}),
+	}, nil
+}
+
+// Execute will add rbac permissions for coredns to sync cluster information
+func (p *updatePhaseCoreDNS) Execute(ctx context.Context) error {
+	_, err := p.kubernetesOperation.Client.RbacV1().ClusterRoles().Create(&rbacv1.ClusterRole{
+		ObjectMeta: metav1.ObjectMeta{Name: corednsResourceName},
+		Rules: []rbacv1.PolicyRule{
+			{
+				APIGroups: []string{""},
+				Verbs:     []string{"list", "watch"},
+				Resources: []string{"endpoints", "services", "namespaces"},
+			},
+		},
+	})
+	err = rigging.ConvertError(err)
+	if err != nil && !trace.IsAlreadyExists(err) {
```

Comment:
For informational/debugging purposes I would log if this resource already exists, otherwise there's 0 feedback in this case.

Reply:
Also, `p.Infof("Created ClusterRole %q", corednsResourceName)` would be nice.